IdentityCore/src/throttling/cache/MSIDLRUCache.m

```diff
@@ -0,0 +1,432 @@
+    }
+    return self;
+}
+
+@end
+
+
+
+//Main class 
+@interface MSIDLRUCache ()
+
+@property (nonatomic) NSUInteger cacheSizeInt;
+@property (nonatomic) NSUInteger cacheUpdateCountInt;
+@property (nonatomic) MSIDLRUCacheNode *head;
+@property (nonatomic) MSIDLRUCacheNode *tail;
+@property (nonatomic) NSMutableDictionary *container;
+@property (nonatomic) NSMutableDictionary *key_signature_map;
+@property (nonatomic) dispatch_queue_t synchronizationQueue;
+
+@end
+
+@implementation MSIDLRUCache
+
+- (NSUInteger)cacheSize
+{
+    return self.cacheSizeInt-2;
+}
+
+- (NSUInteger)numCacheRecords
+{
+    return self.container.allKeys.count-2;
+}
+
+- (NSUInteger)cacheUpdateCount
+{
+    return self.cacheUpdateCountInt;
+}
+
+- (instancetype)initWithCacheSize:(NSUInteger)cacheSize
+{
+    self = [super init];
+    if (self)
+    {
+        _cacheSizeInt = cacheSize+2;
+        _cacheUpdateCountInt = 0;
+        //create dummy head and tail
+        _head = [[MSIDLRUCacheNode alloc] initWithSignature:HEAD_SIGNATURE
+                                              prevSignature:nil
+                                              nextSignature:TAIL_SIGNATURE
+                                                    cacheRecord:nil];
+        
+        _tail = [[MSIDLRUCacheNode alloc] initWithSignature:TAIL_SIGNATURE
+                                              prevSignature:HEAD_SIGNATURE
+                                              nextSignature:nil
+                                                cacheRecord:nil];
+        
+        //create concurrent queue and container dictionary
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.msidlrucache-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+        _container = [NSMutableDictionary new];
+        _key_signature_map = [NSMutableDictionary new];
+        
+        [self.container setObject:_head forKey:HEAD_SIGNATURE];
+        [self.container setObject:_tail forKey:TAIL_SIGNATURE];
+    }
+    return self;
+}
+
++ (MSIDLRUCache *)sharedInstance
+{
+    static MSIDLRUCache *m_service;
+    static dispatch_once_t once_token;
+    
+    dispatch_once(&once_token, ^{
+        m_service = [[MSIDLRUCache alloc] initWithCacheSize:DEFAULT_CACHE_SIZE];
+    });
+    
+    return m_service;
+}
+
+/* add new node to the front of LRU cache.
+if node already exists, update and move it to the front of LRU cache */
+- (BOOL)addToCache:(id)key
+       cacheRecord:(id)cacheRecord
+             error:(NSError *__nullable*__nullable)error
+{
+    __block NSError *subError = nil;
+    BOOL result = YES;
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        if (self.cacheSizeInt <= 2)
+        {
+            subError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"MSIDLRUCache Error: Attempting to write to an empty cache!", nil, nil, nil, nil, nil, YES);
```

Comment:
The error message is a bit confusing. I think you should validate the provided cache size in the init, so it should be >= 1. Then you can remove validation here.

Reply:
Yeah the message is confusing. I will change it to. Cache was initialized as an empty cache. Cannot write to cache of size 0.